internal/usecase/order.go

```diff
@@ -0,0 +1,148 @@
+package usecase
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"gophermart/internal/entity"
+	"gophermart/pkg/logger"
+	"time"
+)
+
+type OrderUseCase struct {
+	repo      OrderRepo
+	repoBonus BonusRepo
+	numAlg    OrderNumAlg
+	accrual   OrderAccrual
+}
+
+func NewOrder(repo OrderRepo, repoBonus BonusRepo, numAlg OrderNumAlg, accrual OrderAccrual) *OrderUseCase {
+	return &OrderUseCase{
+		repo:      repo,
+		repoBonus: repoBonus,
+		numAlg:    numAlg,
+		accrual:   accrual,
+	}
+}
+
+func (uc *OrderUseCase) add(ctx context.Context, order entity.Order) error {
+	logger.Debug().Interface("order", order).Msg("OrderUseCase - add")
+	// Проверяем, что номер заказа допустим
+	ok, err := uc.numAlg.Check(order.Number)
+	if err != nil {
+		return fmt.Errorf("OrderUseCase - Add - uc.numAlg.Check: %w", err)
+	}
+	if !ok {
+		return entity.ErrOrderNumberFormat
+	}
+	// Получаем заказ если он уже есть
+	existsOrder, err := uc.repo.Get(ctx, order.Number)
+	if err != nil {
+		if !errors.Is(err, entity.ErrOrderNotFound) {
+			return fmt.Errorf("OrderUseCase - Add - uc.repo.Get: %w", err)
+		}
+	} else {
+		if existsOrder.UserID != order.UserID {
+			return entity.ErrOrderNumberOwnedByAnotherUser
+		}
+		if existsOrder.UserID == order.UserID {
+			return entity.ErrOrderNumberAlreadyLoaded
+		}
+	}
+	// Добавляем новый номер заказа
+	err = uc.repo.Add(ctx, order)
```

Comment:
нужно разеделить на два запроса. Можно перенести в сервисы
```suggestion
	err = uc.orderRepo.Add(ctx, order)
	err = uc.balanceRepo.Update(newBalance)
```